src/controller.rs

```diff
@@ -335,10 +457,107 @@ async fn launch_job(
+    let job_name = job_name_for(app_instance);
+
+    let pods = pods_api
+        .list(&ListParams {
+            label_selector: Some(format!("job-name={job_name},controller-uid={job_uid}")),
+            ..Default::default()
+        })
+        .await?;
+
+    let mut per_container_logs = HashMap::new();
+
+    // There should be exactly one pod per job. In the unlikely even
+    // something is broken with k8s and we end up getting two pods matching the same job uid
+    // let's just get the logs of all these pods and concatenate them together. Chances are
+    // that this is easier to debug than just getting logs for a random pod.
     //
-    // 1. if job exists check if it's has terminated and take action
-    // 2. make sure we watch the job as it changes status
+    // NOTE(mkm): I don't know how likely this is to happen so I'm not sure if it's worth doing
+    // something more complicated like capturing the pod names and grouping the logs by pod name.
+    for pod in pods.items {
+        let mut container_names = vec![];
+
+        let pod_status = pod.status.as_ref().unwrap();
+        let container_statuses = [
+            pod_status.init_container_statuses.as_ref(),
+            pod_status.container_statuses.as_ref(),
+        ]
+        .into_iter()
+        .flatten()
+        .flat_map(|vec| vec.iter());
+
+        for status in container_statuses {
+            // we cannot get logs from a container that hasn't started yet.
+            // We know a container hasn't started yet when:
+            // 1. the container is explicitly in the "waiting" state
+            // 2. the state field is empty
+            let is_waiting = status
+                .state
+                .as_ref()
+                .map(|x| x.waiting.is_some())
+                .unwrap_or(true);
+            info!(name = status.name, ?is_waiting, "Container status");
+            if !is_waiting {
+                container_names.push(&status.name);
+            }
+        }
+
+        for container_name in container_names {
+            let logs = pods_api
+                .logs(
+                    &pod.name_any(),
+                    &LogParams {
+                        container: Some(container_name.clone()),
+                        ..Default::default()
+                    },
+                )
+                .await?;
+            per_container_logs
+                .entry(container_name.clone())
+                .and_modify(|e: &mut String| e.push_str(&logs))
+                .or_insert(logs);
+        }
+        let logs_json =
+            serde_json::to_string(&per_container_logs).expect("cannot render basic json");
+        info!(logs_json);
+    }
 
+    let app_instance_api: Api<AppInstance> = Api::namespaced(ctx.client.clone(), ns);
+
+    let app_instance_patch = json!({
+        "status": AppInstanceStatus { last_logs: Some(per_container_logs) }
+    });
+    app_instance_api
+        .replace_status(
```

Comment:
Since this [`replace_status`](https://docs.rs/kube/latest/kube/struct.Api.html#method.replace_status) function does a replace on the entire status field, could we perhaps miss something by having this fully altered on the `JobTerminated` `ReconillationState`.

I don't think we do because its only on termination of a `Job` so it only happens in a cleanup, but raising just in case there's something I'm missing too 😄

Reply:
the replace status is done on a app_instance status. We're not storing the state enum in the status, we're computing the status at every reconciliation enumeration by observing what's out there.

that said, I should be using Apply and not Replace